lib/synapse/service_watcher/zookeeper/zookeeper.rb

```diff
@@ -113,38 +113,55 @@ def start_discovery
     end
 
     # find the current backends at the discovery path
-    def discover(zookeeper_opts={:watch => true})
+    def discover(zookeeper_opts={:watch => true}, zk_event=nil)
       statsd_increment('synapse.watcher.zk.discovery', ["zk_cluster:#{@zk_cluster}", "zk_path:#{@discovery['path']}", "service_name:#{@name}"])
       statsd_time('synapse.watcher.zk.discovery.elapsed_time', ["zk_cluster:#{@zk_cluster}", "zk_path:#{@discovery['path']}", "service_name:#{@name}"]) do
         log.info "synapse: discovering backends for service #{@name}"
 
-        new_backends = []
-        zk_children = with_retry(@retry_policy.merge({'retriable_errors' => ZK_RETRIABLE_ERRORS})) do |attempts|
-            statsd_time('synapse.watcher.zk.children.elapsed_time', ["zk_cluster:#{@zk_cluster}", "service_name:#{@name}"]) do
-              log.info "synapse: zk list children at #{@discovery['path']} for #{attempts} times"
-              begin
-                @zk.children(@discovery['path'], zookeeper_opts)
-              rescue ZK::Exceptions::NoNode
-                log.error "synapse: zk list children failed with no node"
-                statsd_increment('synapse.watcher.zk.children.failed', ["zk_cluster:#{@zk_cluster}", "service_name:#{@name}"])
-                []
+        # There are three possible cases, with each case exclusive to others.
+        # 1. No zk event, read data with both data/child watches registered
+        # 2. Registered child watch fired, need_get_children is set to true and children are read from zk with new child watch registered.
+        # 3. Registered data watch fired, need_read_config is set to true and znode is read from zk with new data watch registered.
+        need_get_children = zk_event.nil? || zk_event.node_child?
+        need_read_config = zk_event.nil? || zk_event.node_changed? || zk_event.node_created? || zk_event.node_deleted?
```

Comment:
NIT: under some cases, like `node_deleted`, it is actually not to read the config. Suggest changing it to `need_update_config`.